alphastats/llm/uniprot_utils.py

```diff
@@ -323,7 +346,108 @@ def get_gene_function(gene_name: Union[str, Dict], organism_id=9606) -> str:
     if result and extract_data(result)["functionComments"]:
         return str(extract_data(result)["functionComments"])
     else:
         return "No data found"
+
+
+def get_uniprot_data_for_ids(ids: list):
+    return [get_uniprot_data(protein_id=id)["results"][0] for id in ids]
+
+
+def select_uniprot_id_from_feature(
+    feature: str,
+):
+    """Get uniprot information for a feaure.
+
+    This function collects the results for all base ids (truncating isoform ids) and selects the one to use for feeding the LLM.
+    It does so by reducing the number of results until 1 remains and if that is not straight forward, selects the best annotated (sp over trembl, high annotation score).
+    """
+
+    baseids = sorted(
+        list(set([identifier.split("-")[0] for identifier in feature.split(";")]))
+    )
+    results = get_uniprot_data_for_ids(baseids)
+
+    if len(results) == 1:
+        return results[0]
+
+    # remove inactive entries and ones without gene names (besides immunoglobulins)
+    results = [result for result in results if result["entryType"] != "Inactive"]
+
+    if len(results) == 1:
+        return results[0]
+
+    # remove ones without gene names (besides immunoglobulins)
+    results = [
+        result
+        for result in results
+        if result.get("genes", None) is not None
+        or bool(
+            re.match(
+                ".*globulin.*|^Ig.*",
+                result.get("proteinDescription", {})
+                .get("recommendedName", {})
+                .get("fullName", {})
+                .get("value", ""),
+            )
+        )
+    ]
+
+    if len(results) == 1:
+        return results[0]
+    elif len(results) == 0:
+        return "No data found"
+
+    # Go by gene names, swissprot and annotation scores
+    sp_indices = [
+        i
+        for i, result in enumerate(results)
+        if result["entryType"] == "UniProtKB reviewed (Swiss-Prot)"
+    ]
+    gene_names = [
+        result.get("genes", [{}])[0].get("geneName", {}).get("value")
+        for result in results
+    ]
+    annotation_scores = [result.get("annotationScore") for result in results]
+
+    if (len(set(gene_names)) == 1 and len(sp_indices) > 0) or (
+        len(set(gene_names)) > 1 and len(sp_indices) == 1
+    ):
+        # Either all the same gene name and any swissprot entries, or multiple gene names but only one swissprot entry
+        index = sp_indices[0]
+    elif len(sp_indices) == 0:
+        # Multiple gene names and no swissprot entries present
+        index = annotation_scores.index(max(annotation_scores))
+    else:
+        # Multiple gene names and multiple swissprot entries
+        sp_annotation_scores = [
+            score for i, score in enumerate(annotation_scores) if i in sp_indices
+        ]
+        index = sp_indices[sp_annotation_scores.index(max(sp_annotation_scores))]
+    return results[index]
+
+
+def extract_fieldinformation_from_uniprotresult(
```

Comment:
had to blink a couple of times until i got it x)
```suggestion
def extract_field_information_from_uniprot_result(
```

Reply:
I stand by the beauty of german compound words, but see they have no place in function names. Also this is renamed to entirely to _filter_extracted_annotations in https://github.com/MannLabs/alphapeptstats/pull/383